File: debezium/persistent/PostgresUpsertHandler.java
package com.tiket.tix.hotel.common.debezium.persistent;

import com.tiket.tix.hotel.common.debezium.config.DebeziumConfiguration;
import com.tiket.tix.hotel.common.debezium.DebeziumHelper;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import java.sql.Statement;

public class PostgresUpsertHandler implements UpsertHandler {

  private final DebeziumConfiguration config;
  private final Gson gson = new Gson();

  public PostgresUpsertHandler(DebeziumConfiguration config) {
    this.config = config;
    migratePostgres();
  }

  @Override
  public void upsert(String json) {
    try {
      JsonObject root = gson.fromJson(json, JsonObject.class);
      JsonObject payload = root.getAsJsonObject("payload");
      JsonObject after = payload != null && payload.has("after")
          ? payload.getAsJsonObject("after")
          : null;
      if (after == null) return;

      try (Connection conn = DriverManager.getConnection(
          config.getTargetConfiguration().getPostgresUrl(),
          config.getTargetConfiguration().getPostgresUser(),
          config.getTargetConfiguration().getPostgresPassword())) {

        String sql = "INSERT INTO unified_system_parameter (variable, value, description) " +
            "VALUES (?, ?, ?) " +
            "ON CONFLICT (variable) DO UPDATE SET value = EXCLUDED.value, description = EXCLUDED.description";

        try (PreparedStatement ps = conn.prepareStatement(sql)) {
          ps.setString(1, after.get("variable").getAsString());
          // parse value string JSON jadi string biasa
          ps.setString(2, after.get("value").getAsString());
          ps.setString(3, after.has("description") && !after.get("description").isJsonNull()
              ? after.get("description").getAsString()
              : null);
          ps.executeUpdate();
        }
      }
      DebeziumHelper.sendMetric("postgres_upsert_success");
    } catch (Exception e) {
      DebeziumHelper.sendMetric("postgres_upsert_failed");
      e.printStackTrace();
    }
  }


  private void migratePostgres() {
    try (Connection conn = DriverManager.getConnection(
        config.getTargetConfiguration().getPostgresUrl(),
        config.getTargetConfiguration().getPostgresUser(),
        config.getTargetConfiguration().getPostgresPassword());
        Statement stmt = conn.createStatement()) {

      String sql = "CREATE TABLE IF NOT EXISTS unified_system_parameter (" +
          "variable VARCHAR(255) PRIMARY KEY," +
          "value TEXT NOT NULL," +
          "description TEXT" +
          ")";
      stmt.execute(sql);
      System.out.println("Postgres migration done.");

    } catch (Exception e) {
      e.printStackTrace();
    }
  }

}

File: debezium/persistent/postgres_upsert_handler.go
package persistent

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log"

	_ "github.com/lib/pq"
	"github.com/tiket/TIX-HOTEL-UTILITIES-GO/debezium/config"
	"github.com/tiket/TIX-HOTEL-UTILITIES-GO/debezium/helper"
)

type PostgresUpsertHandler struct {
	cfg *config.DebeziumConfig
	Db  *sql.DB
}

func NewPostgresUpsertHandler(cfg *config.DebeziumConfig) *PostgresUpsertHandler {
	db, err := sql.Open("postgres", cfg.TargetConfig.PostgresURL)
	if err != nil {
		log.Println("Postgres connection failed:", err)
		return nil
	}

	h := &PostgresUpsertHandler{
		cfg: cfg,
		Db:  db,
	}

	h.migratePostgres()
	return h
}

func (h *PostgresUpsertHandler) Upsert(jsonStr string) error {
	var root map[string]interface{}
	if err := json.Unmarshal([]byte(jsonStr), &root); err != nil {
		log.Println("Postgres upsert failed: unmarshal error", err)
		return err
	}

	payload, ok := root["payload"].(map[string]interface{})
	if !ok {
		return nil
	}

	after, ok := payload["after"].(map[string]interface{})
	if !ok || after == nil {
		return nil
	}

	variable := after["variable"].(string)
	value := after["value"].(string)

	var description *string
	if d, ok := after["description"].(string); ok {
		description = &d
	}

	var version *float64
	if v, ok := after["version"].(float64); ok {
		version = &v
	}

	sqlStmt := `
		INSERT INTO unified_system_parameter (variable, value, description, version)
		VALUES ($1, $2, $3, $4)
		ON CONFLICT (variable) DO UPDATE SET value = EXCLUDED.value, description = EXCLUDED.description, version = EXCLUDED.version
	`

	if _, err := h.Db.Exec(sqlStmt, variable, value, description, version); err != nil {
		helper.SendMetric(helper.TagBuilder{
			EventType: "postgres_upsert_failed",
			Variable:  variable,
			Version:   fmt.Sprintf("%.0f", *version),
		})
		return err
	}

	helper.SendMetric(helper.TagBuilder{
		EventType: "postgres_upsert_success",
		Variable:  variable,
		Version:   fmt.Sprintf("%.0f", *version),
	})

	return nil
}

func (h *PostgresUpsertHandler) migratePostgres() {
	sqlStmt := `
	CREATE TABLE IF NOT EXISTS unified_system_parameter (
		variable VARCHAR(255) PRIMARY KEY,
		value TEXT NOT NULL,
		description TEXT,
		version FLOAT
	);

	-- function to create trigger NOTIFY
	CREATE OR REPLACE FUNCTION notify_unified_system_parameter()
	RETURNS trigger AS $$
	BEGIN
		PERFORM pg_notify(
			'unified_system_parameter_channel',
			NEW.variable
		);
		RETURN NEW;
	END;
	$$ LANGUAGE plpgsql;

	-- trigger after insert or update
	DO $$
	BEGIN
		IF NOT EXISTS (
			SELECT 1 FROM pg_trigger WHERE tgname = 'unified_system_parameter_trigger'
		) THEN
			CREATE TRIGGER unified_system_parameter_trigger
			AFTER INSERT OR UPDATE ON unified_system_parameter
			FOR EACH ROW
			EXECUTE FUNCTION notify_unified_system_parameter();
		END IF;
	END;
	$$;
	`

	if _, err := h.Db.Exec(sqlStmt); err != nil {
		log.Println("Postgres migration failed:", err)
		return
	}

	fmt.Println("Postgres migration done.")
}
